requre/storage.py

```diff
@@ -21,14 +21,152 @@
 # SOFTWARE.
 
 import os
+import time
 from _collections_abc import Hashable
 from typing import Dict, List, Any, Optional
 
+from enum import Enum
 import yaml
 
 from .exceptions import PersistentStorageException
 from .singleton import SingletonMeta
-from .constants import VERSION_REQURE_FILE
+from .constants import VERSION_REQURE_FILE, ENV_STORAGE_FILE
+
+# use this sleep to avoid decorating original time function used internally
+original_sleep = time.sleep
+original_time = time.time
+
+
+class DataStructure:
+    """
+    Object model for storing data to persistent storage
+    """
+
+    OUTPUT_KEY = "output"
+    METADATA_KEY = "metadata"
+
+    def __init__(self, output: Any):
+        self.output = output
+        self._metadata: dict = dict()
+
+    @property
+    def metadata(self):
+        """
+        Medatata for item in storage
+        """
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, values: dict):
+        for k, v in values.items():
+            self._metadata[k] = v
+
+    def dump(self):
+        """
+        Dump object to serializable format
+
+        :return dict
+        """
+        return {self.METADATA_KEY: self.metadata, self.OUTPUT_KEY: self.output}
+
+    @classmethod
+    def create_from_dict(cls, dict_repr: dict):
+        """
+        Create Object representation from dict
+
+        :return DataStructure
+        """
+        data = cls(dict_repr[cls.OUTPUT_KEY])
+        data.metadata = dict_repr[cls.METADATA_KEY]
+        return data
+
+
+class DataTypes(Enum):
+    List = 1
+    Dict = 2
+
+
+class DataMiner(metaclass=SingletonMeta):
+    """
+    Intermediate class used for proper formatting and keeping backward
+    compatibility with requre storage version files
+    """
+
+    current_time = original_time()
+    data: DataStructure
+    data_type: DataTypes = DataTypes.List
+    use_latency = False
+    LATENCY_KEY = "latency"
+
+    def get_latency(self, regenerate=True) -> float:
+        """
+        Returns latency from last store event
+
+        :param regenerate: if False, then it will not be changed, just displayed
+        :return: float
+        """
+        current_time = original_time()
+        diff = current_time - self.current_time
+        if regenerate:
+            self.current_time = current_time
+        return diff
+
+    def dump(self, level, key, values):
+        """
+        Store values in proper format to level[key] item
+
+        :param level: parent dict object
+        :param key: item in dict
+        :param values: what to strore (return value of function)
+        :return: None
+        """
+        ds = DataStructure(values)
+        ds.metadata = {self.LATENCY_KEY: self.get_latency()}
+        self.data = ds
+        item = ds.dump()
+        if self.data_type == DataTypes.List:
```

Comment:
We can create classes for each of the types and have some loading/storing methods. (But we can do it later.)

Reply:
Yep, could be done later, I agree with this change